src/dataframe/dataframe.jl

```diff
@@ -1194,14 +1204,38 @@ function Base.append!(df1::DataFrame, df2::AbstractDataFrame)
     end
     ncol(df2) == 0 && return df1
 
-    _names(df1) == _names(df2) || throw(ArgumentError("Column names do not match"))
+    if cols == :orderequal && _names(df1) != _names(df2)
+        if ncol(df1) != ncol(df2)
+            throw(ArgumentError("The numer of columns in both data frames does not match " *
```

Comment:
> Also, this message would also be appropriate if `cols=:setequal`, so it's weird to mention `cols=:orderequal`.

In a corner case of `df2` being a `SubDataFrame` this does not have to be true as `SubDataFrame` can have duplicate columns with the same names which are aliases (see the comment in the code below).

Reply:
So technically if `:orderequal` then `df1` and `df2` can have the same sets of column names, but still have different number of columns (this is what we originally catch in the first condition).

Here is the case I have in mind:
```
julia> df = DataFrame(a=1)
1×1 DataFrame
│ Row │ a     │
│     │ Int64 │
├─────┼───────┤
│ 1   │ 1     │

julia> dfv = @view df[:, [1,1,1,1]]
1×4 SubDataFrame
│ Row │ a     │ a     │ a     │ a     │
│     │ Int64 │ Int64 │ Int64 │ Int64 │
├─────┼───────┼───────┼───────┼───────┤
│ 1   │ 1     │ 1     │ 1     │ 1     │
```

I have rewritten the error message following your suggestion.

Reply:
Why would we ever want column duplication?

Reply:
We do not want it, but we want `DataFrameRow` and `SubDataFrame` constructors to be cheap (this was something we have tested with @nalimilan when `view` cleanup was performed).

We could change it - by adding a test that duplication does not happen (it would add some overhead).
So please comment (+ @nalimilan) if we want to change this (we could possibly add a kwarg that would turn on/off this check).

The good thing is that it is largely harmless - it only means that you get the same name and the same reference to the parent if you pass two indices that point to the same column.

Reply:
Yes, why not add a check for duplicates inside a `@boundscheck` block? Better require things to be simple. Also we could avoid these checks in common cases like `df[i, :]` and `eachrow`. But TBH I don't fully remember the details of that discussion.

Reply:
Though even before/without changing this, couldn't you apply the `:setequal` check to all cases first, and then apply another check for `:orderequal`? It's not a big deal if the error isn't super nice in corner cases like that (especially if we want to forbid them).

Reply:
I think that in `:orderequal` it is more useful what I report now (exact column indices that are problematic - in case when this is possible).

Also in `:orderequal` the test (`==`) is cheaper than `setdiff` in `:setequal` so if someone wants `:orderequal` it will be slightly faster (not much probably though).

Regarding duplicates in column names. `df[r, :]` will be always fast (we use `Index` not `SubIndex` then). The issue was that we wanted to make https://github.com/JuliaData/DataFrames.jl/blob/master/src/other/index.jl#L374 lazy (not to materialize `remap` on creation of `SubDataFrame`).
Currently we do have `@boundscheck` here https://github.com/JuliaData/DataFrames.jl/blob/master/src/other/index.jl#L358 but it does much less work. The reason was to squeeze-out maximum performance of creation of `SubIndex` (some extensive performance tests were run to ensure that this is the case).

But we can call `lazyremap!` inside `@boundscheck` in https://github.com/JuliaData/DataFrames.jl/blob/master/src/other/index.jl#L360 and then we would get what you propose. So - do we go for this?

Reply:
You know much better that code than I do. How large do you think the overhead would be?

Reply:
There are two overheads. First - in the error check we are discussing here. The difference is:
```
julia> x = Symbol.("x", 1:50);

julia> y = Symbol.("x", 1:50);

julia> using BenchmarkTools

julia> @benchmark $x == $y
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     34.339 ns (0.00% GC)
  median time:      34.642 ns (0.00% GC)
  mean time:        37.062 ns (0.00% GC)
  maximum time:     265.257 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     993

julia> f(x,y) = symdiff(Set(x), Set(y))
f (generic function with 1 method)

julia> @benchmark f($x, $y)
BenchmarkTools.Trial:
  memory estimate:  11.22 KiB
  allocs estimate:  33
  --------------
  minimum time:     5.820 μs (0.00% GC)
  median time:      6.400 μs (0.00% GC)
  mean time:        10.529 μs (17.63% GC)
  maximum time:     7.190 ms (99.80% GC)
  --------------
  samples:          10000
  evals/sample:     5
```
So the overhead of `:setequal` over currrent `:orderequal` is on the average several μs.

Now the check for deduplication will have the following cost:
```
julia> x = [1:50;];

julia> @benchmark allunique($x)
BenchmarkTools.Trial:
  memory estimate:  3.88 KiB
  allocs estimate:  10
  --------------
  minimum time:     1.100 μs (0.00% GC)
  median time:      1.190 μs (0.00% GC)
  mean time:        2.587 μs (29.39% GC)
  maximum time:     4.025 ms (99.94% GC)
  --------------
  samples:          10000
  evals/sample:     10
```
So it is around one μs.

I have chosen 50 columns in both tests as I think this is a typical maximum.

So in summary: in both cases we are talking about the overhead is of μs.

Now as a benchmark - currently the cost of creating a `DataFrameRow` is:
```
julia> @benchmark $df[1, $x]
BenchmarkTools.Trial:
  memory estimate:  144 bytes
  allocs estimate:  3
  --------------
  minimum time:     69.743 ns (0.00% GC)
  median time:      73.027 ns (0.00% GC)
  mean time:        99.331 ns (14.20% GC)
  maximum time:     41.290 μs (99.69% GC)
  --------------
  samples:          10000
  evals/sample:     975
```
and `SubDataFrame` is:
```
julia> @benchmark $df[[1], $x]
BenchmarkTools.Trial:
  memory estimate:  23.38 KiB
  allocs estimate:  155
  --------------
  minimum time:     11.499 μs (0.00% GC)
  median time:      12.800 μs (0.00% GC)
  mean time:        20.584 μs (30.09% GC)
  maximum time:     41.230 ms (99.90% GC)
  --------------
  samples:          10000
  evals/sample:     1
```

(all the time we have 50 columns)

So in short - for `DataFrameRow` it will be noticeable (an order of magnitude slower), but for `SubDataFrame` it is probably + 10-20% of time.

Finally note that if we use `:` to create `DataFrameRow` or `SubDataFrame` this cost would not be incurred (and this is probably the most common case).

So - given this evidence - what do you think we should do?

For me:
* I would leave error checks as is (to make `:orderequal` as fast as possible)
* for checking `DataFrameRow` and `SubDataFrame` for duplicate columns - I am not sure. The cost is non negligible, but on the other hand it is not huge. Also checking for duplicate columns would simplyfy `SubIndex` code (no need to remap lazily - we would just remap).

Reply:
Though not being meant, here are my two cents.
* As I will rewrite/reorder the whole error checking (as part of the introduction of more `cols` keywords) anyway, I would just keep it the way it is right now (even though that might increase the speed regression that eventually results from my rewrite).
* I would prefer some contracted check here, likely by making the duplicate check a boundscheck and whenever people really seek for the full performance they may declare @inbounds to guarantee - by contract - that they don't utilize duplicate columns. Otherwise it'd plain check it and would try to optimize every internal call such that we don't get the penalty for internal usage (when we guarantee that everything won't have duplicate columns) -> have some sort of barrier such that internal usage will assume no duplicates. I hope that was a feasible explanation of my thoughts.